src/boot/fwupd.c

```diff
@@ -0,0 +1,589 @@
+        EFI_TIME now;
+        EFI_STATUS err;
+
+        err = RT->GetTime(&now, NULL);
+        if (err != EFI_SUCCESS) /* Should not happen, but sometimes clocks are flaky */
+                now = (EFI_TIME) { };
+
+        FOREACH_ARRAY(u, updates, n_updates) {
+                log_debug("Marking UEFI update variable '%ls' as attempted", (*u)->name);
+
+                        (*u)->info->time_attempted = now;
+                (*u)->info->status = FWUPDATE_ATTEMPTED;
+
+                err = efivar_set_raw(MAKE_GUID_PTR(FWUPDATE), (*u)->name, (*u)->info, (*u)->size, (*u)->attrs);
+                if (err != EFI_SUCCESS)
+                        return log_error_status(err, "Failed to set variable '%ls': %m", (*u)->name);
+        }
+
+        return EFI_SUCCESS;
+}
+
+/* Implement the flow defined at https://github.com/fwupd/fwupd-efi/blob/main/SECURITY.md */
+void fwupd_install_updates(EFI_HANDLE image) {
+        EFI_STATUS err;
+
+        /* Userspace prepares a set of 'fwupd-*' variables pointing to the capsules to install, parse them
+         * and store them in a table. */
+        FwupUpdateTable **updates = NULL;
+        size_t n_updates = 0;
+        CLEANUP_ARRAY(updates, n_updates, fwup_update_table_free_many);
+        err = parse_update_vars(&updates, &n_updates);
+        if (err != EFI_SUCCESS || n_updates == 0)
+                return;
+
+        /* Now prepare the capsules by loading them in memory. This is best-effort: if there are multiple
+         * capsules queued for installation, but some cannot be loaded for any reason, skip them and install
+         * what we can. */
+        _cleanup_free_ EFI_CAPSULE_BLOCK_DESCRIPTOR *capsule_blocks = NULL;
+        EFI_CAPSULE_HEADER **capsules = NULL;
+        size_t n_capsules = 0;
+        CLEANUP_ARRAY(capsules, n_capsules, capsule_free_many);
+        err = prepare_capsules(image, updates, n_updates, &capsules, &n_capsules, &capsule_blocks);
+        if (err != EFI_SUCCESS)
+                return;
+
+        /* Mark all parsed capsules as "attempted" to be updated. This also includes the ones that could not
+         * be loaded, otherwise we would just loop forever. */
+        err = mark_capsules_as_attempted(updates, n_updates);
+        if (err != EFI_SUCCESS)
+                return;
+
+        /* If no capsules could be prepared, log and bail out, no need to reset.
+         * NB: this check is placed after the efi var update, since not all efi vars might reference
+         * existing or valid capsules, but the variables need to be marked as processed anyway in order
+         * to avoid getting stuck in a loop. */
+        if (n_capsules == 0)
+                return (void) log_error("No capsules could be prepared out of %zu updates.", n_updates);
+
+        log_info("Applying %zu updates, and then resetting system.", n_capsules);
+        log_wait();
+
+        EFI_RESET_TYPE reset_type = EfiResetWarm;
```

Comment:
i'd drop the initialization here, since you have two places otherwise where you encode the EfiResetWarm fallbcak